waveform_editor/dependency_graph.py

```diff
@@ -0,0 +1,65 @@
+from waveform_editor.derived_waveform import DerivedWaveform
+
+
+class DependencyGraph:
+    def __init__(self):
+        self.graph = {}
+
+    def add_node(self, name, dependencies):
+        self.graph[name] = set(dependencies)
+        try:
+            self.detect_cycles()
+        except RuntimeError as e:
+            del self.graph[name]
+            raise e
```

Comment:
```suggestion
        except RuntimeError:
            del self.graph[name]
            raise
```

Otherwise the stacktrace is a bit weird:
```python
>>> try:
...     raise RuntimeError("test")
... except RuntimeError:
...     raise
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
RuntimeError: test
>>> try:
...     raise RuntimeError("test")
... except RuntimeError as e:
...     raise e
... 
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
  File "<stdin>", line 2, in <module>
RuntimeError: test
```